### PR TITLE
feat: ip allowlist crud endpoints

### LINKED ISSUE #59: Implement CRUD endpoints for ip allowlist
Build crud endpoints for managing tenant's IP allowlist in the system

```ts
// object id is omitted but necessary

interface IPEntry {
  ip: string;
  comment: string;
}
```

### GET

Returns a list of all allowed IP for a tenant.

### POST

Add a new allowed IP for a tenant. No duplicates allowed.

### DELETE

Delete an IP address from the allowed list.

### PR DESCRIPTION
**Pull request by AnthonyLaiRH**
### Related Issue(s)
  
resolves #59 

  
### Quick overview of the change
CRUD endpoints for IP allowlist.   

Also added the `prisma db push` command to sync with mongodb on `@unique` attributes / indices
More here [prisma release 3.4.0 notes](https://github.com/prisma/prisma/releases/tag/3.4.0)

### CODE REVIEW

File: server/package.json [hunk 1]
```diff
@@ -22,7 +22,8 @@
     "test:e2e": "jest --config ./test/jest-e2e.json",
     "prisma:generate": "npx prisma generate",
     "prisma:generate-watch": "npx prisma generate --watch",
-    "prisma:studio": "npx prisma studio"
+    "prisma:studio": "npx prisma studio",
+    "prisma:db-push": "npx prisma db push"
```
Comment (AnthonyLaiRH) [inline server/package.json:26]: For syncing mongodb with `@unique` attributes 

https://github.com/prisma/prisma/releases/tag/3.4.0

Comment (andrewpratheepan) [reply server/package.json:26]: I won't approve this PR until the clock strikes 1pm

File: server/package.json [hunk 2]
```diff
@@ -13,7 +13,7 @@
     "start:dev": "nest start --watch",
     "start:debug": "nest start --debug --watch",
     "start:prod": "node dist/main",
-    "start:docker": "concurrently --kill-others \"yarn run start:debug\" \"yarn run prisma:generate-watch\" \"yarn run prisma:studio\"",
+    "start:docker": "concurrently --kill-others \"yarn run start:debug\" \"yarn run prisma:generate-watch\" \"yarn run prisma:db-push\" \"yarn run prisma:studio\"",
```
Comment (andrewpratheepan) [inline server/package.json:16]: I get this when I tried to build docker:
```
chronicle-api              | [2] 
chronicle-api              | [2] The database is already in sync with the Prisma schema.
chronicle-api              | [1] 
chronicle-api              | [1] Watching... /usr/src/app/prisma/schema.prisma
chronicle-api              | [1] 
chronicle-api              | [1] ✔ Generated Prisma Client (3.10.0 | library) to ./node_modules/@prisma/client in
chronicle-api              | [1]  257ms
✔ Generated Prisma Client (3.10.0 | library) to ./node_modules/@prisma/client in
chronicle-api              | [2]  196ms
chronicle-api              | [2] 
mongodb                    | {"t":{"$date":"2022-03-06T17:36:10.370+00:00"},"s":"I",  "c":"STORAGE",  "id":22430,   "ctx":"Checkpointer","msg":"WiredTiger message","attr":{"message":"[1646588170:370488][1:0x7facfca40700], WT_SESSION.checkpoint: [WT_VERB_CHECKPOINT_PROGRESS] saving checkpoint snapshot min: 10, snapshot max: 10 snapshot count: 0, oldest timestamp: (0, 0) , meta checkpoint timestamp: (0, 0) base write gen: 7"}}
chronicle-api              | [2] Done in 5.25s.
chronicle-api              | [2] yarn run prisma:db-push exited with code 0
chronicle-api              | --> Sending SIGTERM to other processes..
chronicle-api              | [0] yarn run start:debug exited with code SIGTERM
chronicle-api              | --> Sending SIGTERM to other processes..
chronicle-api              | [3] yarn run prisma:studio exited with code SIGTERM
chronicle-api              | --> Sending SIGTERM to other processes..
chronicle-api              | [1] yarn run prisma:generate-watch exited with code SIGTERM
chronicle-api exited with code 1
```

the `prisma db push` may need to be a part of the [docker step instead](https://github.com/KonomiAI/Chronicle/blob/12cf92f87a08b40847c19f323688a428b5e9386b/server/Dockerfile#L11-L12)

Comment (AnthonyLaiRH) [reply server/package.json:16]: gotchu, im gonna make this into another issue and remove `prisma:db-push` from the docker command for now

File: server/src/models/ip/ip.dto.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  IsString,
+  IsIP,
+} from 'class-validator';
+
+export class AddIPDto {
+  @IsIP()
+  @IsNotEmpty()
+  ip: string;
+
+  @IsNotEmpty()
```
Comment (andrewpratheepan) [inline server/src/models/ip/ip.dto.ts:12]: Should we force `description` to be a part of the request.

File: server/src/models/ip/ip.dto.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  @IsString()
+  description: string;
+}
+
+export class UpdateIPDescriptionDto {
+  @IsNotEmpty()
+  @IsString()
+  description: string;
+}
```
Comment (andrewpratheepan) [inline server/src/models/ip/ip.dto.ts:21]: If we are giving the user to update the description, would it make sense to give them the ability to update the whitelist as well (so if there is a mistake/update, they do not need to create a new entry).

**Review by andrewpratheepan — APPROVED** [review]
Everything looks good so far, just a couple minor comments but should be good to merge.

File: server/src/models/ip/ip.dto.ts (hunk 3 above)
Comment (AnthonyLaiRH) [reply server/src/models/ip/ip.dto.ts:12]: yes! good catch

File: server/src/models/ip/ip.dto.ts (hunk 4 above)
Comment (AnthonyLaiRH) [reply server/src/models/ip/ip.dto.ts:21]: I was actually thinking about that too. My idea for the flow was if the IP was incorrect, a new entry was needed and the incorrect one would need to be deleted.

I can make a quick edit and test out that flow instead, but we can talk about this more

Comment (AnthonyLaiRH) [reply server/src/models/ip/ip.dto.ts:21]: mongo atlas allowlist lets you edit the IP so ill probably change IP to be included in the update

**Pull request merged**